### ISSUE TITLE
tests power iteration methods by comparing against np.linalg.eig results 

### ISSUE THREAD

**Issue by Tron-x** [noahgolmant/pytorch-hessian-eigenthings#22]
![image](https://user-images.githubusercontent.com/7659413/49950853-91d6f980-ff33-11e8-9dc6-f1ed103a86f9.png)
I run the test file which tests the accuracy of the power iteration methods by comparing
against np.linalg.eig results for various random matrix configurations
er....the result seems unstable ,and the error seems too big on some matrix, I'm curious what your observations are. when run the test ,i just run python power_iter_tests.py without other configurations.

**Comment by noahgolmant** [comment]
If you change different hyperparameters (e.g. number of iterations, error threshold) you can reduce this error significantly. The default configuration is just faster. When I change these parameters I can achieve lower than 1% error.

**Comment by Tron-x** [comment]
i change the hyperparameters number of iterations from 20 to 100,error threshold from e-04 to e-05 
![image](https://user-images.githubusercontent.com/7659413/50040757-9ced6f00-0084-11e9-801e-ac749d5b0b62.png)
the error of some matrix still little higher ,seems unstable, I'm curious the reason is?

**Comment by noahgolmant** [comment]
It would be more accurate to average those errors over all the matrices to get the correct test performance. Each printout is a particular random wishart matrix. It's more informative to take the mean and variance of those error values. You should probably expect the variance to increase as the matrix dimensionality increases.

The performance should steadily improve as you increase the number of iterations / error threshold, and momentum accelerates it significantly.

**Comment by Tron-x** [comment]
you are right when i turn number of iterations  to 10000 ,error threshold  e-05
![image](https://user-images.githubusercontent.com/7659413/50042095-0e83e800-009a-11e9-900c-d40933f4b1ad.png)
but when i keep up  turn error threshold to e-06 ,iteration is still 10000,the result seems not good
![image](https://user-images.githubusercontent.com/7659413/50042109-4db23900-009a-11e9-8ba7-ba9a9563fca5.png)
it seems that the threshold  unuseful

**Comment by noahgolmant** [comment]
What do you mean by not good? It looks like there are only two examples that achieve over 5% error. The performance should be taken to be an average over the error of these specific tests

**Comment by Tron-x** [comment]
sorry,I misread the result，several more tests were done.The result mostly looks good. So what do you think are the reasons for some bad error?Does it mean that individual eigenvalues are not calculated correctly?

**Comment by noahgolmant** [comment]
No, it's my fault for not outputting the mean/variance of the tests to begin with. There will be some outliers because the procedure is stochastic. The error should be averaged over something like 30 tests. In those outliers, the eigenvalues are off by a factor of ~2, which can be good or bad depending on the magnitude/variance of the actual matrix spectrum.

**Comment by Tron-x** [comment]
when you call deflated_power_iteration() to compute the hessian value ,you use HVPOperator,it seems every 'step for' you apply.prepare_grad() ,next(self.dataloader_iter) new data inputs ,if i want to compute some moment hessian eigenvalue such  as the 100th iteration when training ,as your method when computing this time point eigenvalue, i need use other batch data(you call it stochastic power iteration )，Is this reasonable? My purpose is to verify “Hessian-based Analysis of Large Batch Training and Robustness to Adversaries" figure 1

**Comment by noahgolmant** [comment]
Figure 1 in that paper computes the spectrum for the final models obtained at the end of training for various SGD batch sizes. To replicate the figure, just train several models using different SGD mini-batch sizes, then call `compute_hessian_eigenthings` on each model.

If you set the batch size parameter for stochastic power iteration to be the size of the training set, you should be able to exactly reproduce that figure since it would no longer be a stochastic estimate. You can see how they do this for a single eigenvalue in their HessianFlow repo that I link to in the acknowledgments section.

A figure that uses a small power iteration batch size might not look the exact same due to the variance in the eigenvalue estimates.

It might be the case that running power iteration on a single mini-batch produces a good estimate of the spectrum across the whole dataset, but this is different from stochastic power iteration, which uses a different batch at each step. I would need to test this out.

**Comment by Tron-x** [comment]
thank you ,but when i run the main.py ,the result is different when use gpu and not use
not use:
![image](https://user-images.githubusercontent.com/7659413/50202808-480a6b00-039a-11e9-882b-663b74b278fb.png)
use gpu:

![image](https://user-images.githubusercontent.com/7659413/50202882-90298d80-039a-11e9-8905-33c6c3e8c0bf.png)
I'm curious what your observations are?

**Comment by noahgolmant** [comment]
The random seed may differ between subsequent runs if you don't set the random seed, and even if you set it, the behavior may differ between the GPU and CPU for torch (see this issue https://discuss.pytorch.org/t/are-gpu-and-cpu-random-seeds-independent/142)

**Comment by Tron-x** [comment]
it seems not as this, when i use cpu only ,the first random vector is 

![image](https://user-images.githubusercontent.com/7659413/50370170-dd953d00-05dc-11e9-9ef0-36d82fc2f705.png)
when i use gpu only the first random vector is 

![image](https://user-images.githubusercontent.com/7659413/50370199-51cfe080-05dd-11e9-8304-27e8af4d9052.png)
the next "n for iteration" the vector of cpu and gpu is same too,
but the result the different(the gap between gpu and cpu is so huge)

![image](https://user-images.githubusercontent.com/7659413/50370263-99a33780-05de-11e9-9c93-025746fb36a0.png)

![image](https://user-images.githubusercontent.com/7659413/50370278-d0794d80-05de-11e9-986c-74b9982e0f09.png)
 it seem that the reason i can think out is caused by the grad_matrix is different(stochastic power iteration),
the grad_matrix is always changing ()，In principle, the results should not be so strongly dependent on randomness（cpu and gpu）. The algorithm(stochastic power iteration) seems to estimate the eigenvalues of global samples by using the eigenvalues of subsamples(many batchs same to iteration times), as a result, the algorithm is not robust(Which result (gpu and cpu)is the right one?)

**Comment by noahgolmant** [comment]
Have you tried changing the batch size / number of steps / momentum term? I am planning to add some tests for this in #18 based on issue #17. This is a procedure proposed in "[Accelerated Stochastic Power Iteration](https://arxiv.org/abs/1707.02670)" and the convergence depends on the variance of the Hessian estimate. So larger networks may require larger batch sizes. In Theorem 3 of this paper, the convergence condition (for our purposes) is given in terms of the spectral norm of the covariance of the Hessian estimate, so this is to be expected.

I would only be concerned if this fails when the batch size is set to the size of the dataloader, since that is vanilla power iteration which should be stable.

**Comment by Tron-x** [comment]
when i run this code [https://github.com/amirgholami/HessianFlow](url)(just can compute top-one),every eporch finished, then compute the eigenvalue and eigen-vector, and call your compute_hessian_eigenthings()，but the result of top-one eigenvalue and eigen-vector by your method is very different with the method of [https://github.com/amirgholami/HessianFlow](url), the two method have a big gap, did you have a test, I'm curious what your observations are, when i fix the random vector,the two methods still have a huge gap.

**Comment by noahgolmant** [comment]
Are you using a mini-batch smaller than the dataset size? I will run some tests very soon to see if this issue persists when I use the full dataset.

I am also going to try some new techniques to reduce the variance of the eigenvalue estimates. This instability is definitely an important issue even if it makes sense theoretically so I'm going to try a few different techniques to fix it up (I've referenced these in #23)

**Comment by Tron-x** [comment]
I don't understand why we need to compute eigenvalues on test sets. I tested that if we iterate over load mini-batch data on full-test-data every time, the two methods (get_eigen_full_dataset() and yours) are still in one order of magnitude although there is a gap between them. But if only eigenvalues are computed on a fixed mini-batch data set, the difference between the two methods (get_eigen() and yours) is large(not in one order of magnitude). I also tested the accuracy of power-iteration in (Hessian Flow) using matlab, which is correct, Can you put two pieces of code together (Hessian Flow and yours) and test them? I'm not so sure about my test. I think it's meaningful to perfect this  calculation. thank you~

**Comment by noahgolmant** [comment]
I am testing this in #23, but I have tested the accuracy of power iteration on both random matrices and the full Hessian my comparing to the `np.linalg.eig` results. The full dataset eigenvalue estimates appear to be fine.

The difference between the methods for the mini-batching technique is that I load a new mini-batch every time (so the gradient is stochastic), whereas Hessian Flow performs power iteration on a single mini-batch. The first way should produce an unbiased estimate of the full Hessian, but it displays high variance.
